src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,154 @@
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState).ConfigureAwait(false);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+
+            return entry;
+        }
+
+        // No return value. Modifies InstanceCache directly.
+        private static async Task DiscoverAsync(string host, bool validateAuthority, CallState callState)
+        {
+            string tentativeAuthorizeEndpoint = AuthorizeEndpointTemplate.Replace("{host}", host);
+            string instanceDiscoveryEndpoint =
+                ("https://{host}/common/discovery/instance?api-version=1.1&authorization_endpoint=" +
+                 tentativeAuthorizeEndpoint)
+                .Replace("{host}", WhitelistedAuthorities.Contains(host) ? host : DefaultTrustedAuthority);
+            var client = new AdalHttpClient(instanceDiscoveryEndpoint, callState);
+            InstanceDiscoveryResponse discoveryResponse = null;
+            try
+            {
+                discoveryResponse = await client.GetResponseAsync<InstanceDiscoveryResponse>().ConfigureAwait(false);
+                if (validateAuthority && discoveryResponse.TenantDiscoveryEndpoint == null)
+                {
+                    // hard stop here
+                    throw new AdalException(AdalError.AuthorityNotInValidList);
+                }
+            }
+            catch (AdalServiceException ex)
+            {
+                if (validateAuthority)
+                {
+                    // hard stop here
+                    throw new AdalException(
+                        (ex.ErrorCode == "invalid_instance")
+                            ? AdalError.AuthorityNotInValidList
+                            : AdalError.AuthorityValidationFailed, ex);
+                }
+            }
+
+            foreach (var entry in discoveryResponse?.Metadata ?? Enumerable.Empty<InstanceDiscoveryMetadataEntry>())
+            {
+                foreach (var alias in entry?.Aliases ?? Enumerable.Empty<string>())
+                {
+                    InstanceCache.TryAdd(alias, entry);
+                }
+            }
+
+            InstanceCache.TryAdd(host, new InstanceDiscoveryMetadataEntry
+            {
+                PreferredNetwork = host,
+                PreferredCache = host,
+                Aliases = null
```

Comment:
I'm not sure you need to set Aliases to null here explicitly -- won't it be by default?

Reply:
It would probably be. It is just that my C habit of being-explicit occassionally comes back.